ampd/src/asyncutil/future.rs

```diff
@@ -0,0 +1,175 @@
+use futures::{Future, FutureExt};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use tokio::time::sleep;
+
+pub fn with_retry<F, Fut, R, Err>(
+    get_future: F,
+    policy: RetryPolicy,
+) -> impl Future<Output = Result<R, Err>>
+where
+    F: Fn() -> Fut,
+    Fut: Future<Output = Result<R, Err>>,
+{
+    FutureRetry::new(get_future, policy)
+}
+
+pub enum RetryPolicy {
+    RepeatConstant(Duration, u64),
+}
+
+struct FutureRetry<F, Fut, R, Err>
+where
+    F: Fn() -> Fut,
+    Fut: Future<Output = Result<R, Err>>,
+{
+    get_future: F,
+    future: Pin<Box<Fut>>,
+    policy: RetryPolicy,
+    err_count: u64,
+}
+
+impl<F, Fut, R, Err> Unpin for FutureRetry<F, Fut, R, Err>
+where
+    F: Fn() -> Fut,
+    Fut: Future<Output = Result<R, Err>>,
+{
+}
```

Comment:
why is this necessary?

Reply:
Let me try to explain this based. on my understanding. At lines [62](https://github.com/axelarnetwork/axelar-amplifier/blob/7e792b785eb713fd04eb761cadd4f964e0109dab/ampd/src/asyncutil/future.rs#L62), [73](https://github.com/axelarnetwork/axelar-amplifier/blob/7e792b785eb713fd04eb761cadd4f964e0109dab/ampd/src/asyncutil/future.rs#L73), and [94](https://github.com/axelarnetwork/axelar-amplifier/blob/7e792b785eb713fd04eb761cadd4f964e0109dab/ampd/src/asyncutil/future.rs#L94), I am trying to mutate the future which is of type `Pin<&mut Self>`.